apps/web/src/components/WalletConnect/SessionProposalModal.tsx

```diff
@@ -0,0 +1,192 @@
+import { CheckmarkIcon, CloseIcon } from "../../assets/icons";
+import { OwnedImplicitAccountsAutocomplete } from "../AddressAutocomplete";
+
+/**
+ * Finds the overlap between the chains supported by the wallet and the chains requested by the dApp.
+ * We take the first required chain as Umami supports connection for only one chain per request.
+ */
+function getChainToConfirm(required: (string | string[] | undefined)[]) {
+  // take the first required chain which is supported by the wallet
+  for (const chain of required.flat()) {
+    if (chain && ["tezos:mainnet", "tezos:ghostnet"].includes(chain)) {
+      return chain;
+    }
+  }
+  return undefined;
+}
+
+export default function SessionProposalModal({
+  proposal,
+}: {
+  proposal: WalletKitTypes.SessionProposal;
+}) {
+  console.log("SessionProposalModal", proposal);
+  const getAccount = useGetImplicitAccount();
+  const toast = useToast();
+
+  const { onClose } = useDynamicModalContext();
+  const { handleAsyncAction } = useAsyncActionHandler();
+
+  const [isLoadingApprove, setIsLoadingApprove] = useState(false);
+  const [isLoadingReject, setIsLoadingReject] = useState(false);
+  const form = useForm<{ address: string }>({
+    mode: "onBlur",
+  });
+  const {
+    getValues,
+    formState: { errors, isValid },
+  } = form;
+
+  const required = Object.entries(proposal.params.requiredNamespaces).map(([key, values]) =>
+    key.includes(":") ? key : values.chains
+  );
+
+  if (required.length !== 1) {
+    console.error("Expected only one required chain", required);
+  }
+
+  const chain = getChainToConfirm(required);
+  if (!chain) {
+    console.error("No supported chain found", required);
+  }
+
+  const onApprove = () =>
+    handleAsyncAction(async () => {
+      const account = getAccount(getValues().address);
+
+      try {
+        if (!chain) {
+          throw new Error("No supported chain found");
+        }
+        const namespaces = buildApprovedNamespaces({
+          proposal: proposal.params,
+          supportedNamespaces: {
+            tezos: {
+              chains: [chain],
+              methods: ["tezos_getAccounts", "tezos_sign", "tezos_send"],
+              events: [],
+              accounts: [`${chain}:${account.address.pkh}`],
+            },
+          },
+        });
+        console.log("approvedNamespaces", namespaces);
+
+        setIsLoadingApprove(true);
+        await walletKit.approveSession({
+          id: proposal.id,
+          namespaces,
+          sessionProperties: {},
+        });
+        onClose();
+      } catch (e) {
+        setIsLoadingApprove(false);
+        toast({ description: (e as Error).message, status: "error" });
+        return;
+      }
+
+      setIsLoadingApprove(false);
+    }).finally(onClose);
+
+  // Handle reject action
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const onReject = () =>
+    handleAsyncAction(async () => {
+      try {
+        setIsLoadingReject(true);
```

Comment:
`setIsLoadingReject` could be moved to `finally` block after `catch` inside `handleAsyncAction`
the same could be applied to the approve logic